content/advent-2015/goaUntanglingMicroservices.md

```diff
@@ -0,0 +1,333 @@
+ -h, --host="cellar.goa.design"  
+		      API hostname
+ -t, --timeout=20s    Set the request timeout, defaults to 20s
+     --dump           Dump HTTP request and response.
+     --pp             Pretty print response body
+
+Commands:
+ help [<command>...]
+   Show help.
+
+
+ create bottle [<flags>] <path>
+   Record new bottle
+
+   --payload=PAYLOAD  Request JSON body
+
+ delete bottle <path>
+
+ list bottle [<flags>] <path>
+   List all bottles in account optionally filtering by year
+
+   --years=YEARS  Filter by years
+
+ show bottle <path>
+   Retrieve bottle with given id
+```
+The tool also provides contextual help for each action:
+```
+./cellar-cli show bottle --help
+usage: cellar-cli show bottle <path>
+
+Retrieve bottle with given id
+
+Flags:
+      --help           Show context-sensitive help (also try --help-long and
+                       --help-man).
+  -s, --scheme="http"  Set the requests scheme
+  -h, --host="cellar.goa.design"  
+                       API hostname
+  -t, --timeout=20s    Set the request timeout, defaults to 20s
+      --dump           Dump HTTP request and response.
+      --pp             Pretty print response body
+
+Args:
+  <path>  Request path, format is /cellar/accounts/:accountID/bottles/:bottleID
+```
+The implementation of the client tool relies on the generated client
+Go package to make the actual requests. The package exposes a function
+for each action exposed by the API, see the files generated for the
+cellar example [clients](https://github.com/raphael/goa/blob/master/examples/cellar/client)
+for more details.
+
+The other client `goagen` can produce is the JavaScript module. The
+module can be used by both client and server side JavaScript.
+Similarly to the Go package the JavaScript module exposes one function
+per API action. It uses the [axios](https://github.com/mzabriskie/axios)
+library to make the actual requests. Again the cellar example contains
+the [generated JavaScript](https://github.com/raphael/goa/blob/master/examples/cellar/js/client.js)
+together with [an example](https://github.com/raphael/goa/blob/master/examples/cellar/js/index.html)
+on how to use it if you are curious.
+
+Summing it all up, the diagram below shows all the various outputs of
+the `goagen` tool:
+![goagen diagram](https://cdn.rawgit.com/raphael/goa/master/images/goagenv4.svg "goagen")
+
+## The Engine: Runtime
+
+goa is not just about code generation though. The package also provides
+a powerful *context* object that makes it possible to carry deadlines
+and cancellation signals to all the request handlers. The
+[Timeout middleware](https://godoc.org/github.com/raphael/goa#Timeout)
+takes advantage of that to send a cancelation signal to the request
+handler after a given amount of time. As we've seen before, the context
+object also exposes the request and response states wrapping them in
+convenient methods that are specific to each action as described in the
+design.
+
```

Comment:
Expand this?  {since the context is carried through every layer of your application you have access to all the state specific to each request from every corner of your application } What does this mean for logging, for data access, auth, etc?